herddb-core/src/test/java/herddb/core/CreateTableTest.java

```diff
@@ -105,8 +106,67 @@ public void testWithNullPrimaryKey2() throws Exception {
                     .primaryKey("id")
                     .build();
 
+            CreateTableStatement st2 = new CreateTableStatement(table);
+
+            manager.executeStatement(st2, StatementEvaluationContext.DEFAULT_EVALUATION_CONTEXT(), TransactionContext.NO_TRANSACTION);
+        }
+    }
+
+
+    @Test(expected = StatementExecutionException.class)
+    public void weThrowExceptionOnInsertingNullInNonNullColumn() throws Exception {
+        String nodeId = "localhost";
+        try (DBManager manager = new DBManager("localhost", new MemoryMetadataStorageManager(), new MemoryDataStorageManager(), new MemoryCommitLogManager(),null, null);) {
+            manager.start();
+            CreateTableSpaceStatement st1 = new CreateTableSpaceStatement("tblspace1", Collections.singleton(nodeId), nodeId, 1, 0, 0);
+            manager.executeStatement(st1, StatementEvaluationContext.DEFAULT_EVALUATION_CONTEXT(), TransactionContext.NO_TRANSACTION);
+            manager.waitForTablespace("tblspace1", 10000);
+
+            Table table = Table
+                    .builder()
+                    .tablespace("tblspace1")
+                    .name("t1")
+                    .column("id", ColumnTypes.STRING)
+                    .column("name", ColumnTypes.STRING)
+                    .primaryKey("id")
+                    .build();
+
             CreateTableStatement st2 = new CreateTableStatement(table);
             manager.executeStatement(st2, StatementEvaluationContext.DEFAULT_EVALUATION_CONTEXT(), TransactionContext.NO_TRANSACTION);
+            TranslatedQuery translated = manager.getPlanner().translate("tblspace1", "INSERT INTO t1 (id,name) values(?,?)", Arrays.asList(null, 1234), true, true, false, -1);
+            manager.executePlan(translated.plan, translated.context, TransactionContext.NO_TRANSACTION);
         }
     }
+
```

Comment:
The reason for this UT to be commented out is -
1. I have some observations here. The null check on column only works if they are part of the primary key. How can we enforce the same if non primary key columns have data types which are not null? 
2. I don't see an easy way to add a test for update query with null value for the primary key. It just ignore it and keeps the original record.

Reply:
Actually you cannot update the PK (it is better to file an issue).
It would be like a DELETE + INSERT, but it is a case not actually covered.

> The null check on column only works if they are part of the primary key
I am expecting that with your patch any NULL in a NOT-NULL column must fail the serialization of the record, and this will lead to an error to the client.

If this is not happening check with a break point here 
https://github.com/diennea/herddb/blob/6393743bafb800c7e86fdd86b52019d1903a0bb2/herddb-core/src/main/java/herddb/core/TableManager.java#L964

We are catching any "StatementExecutionException" and rethrowing it as a validation error, maybe you have to throw a StatementExecutionException during the serialization of the record

Reply:
This is where you have to put the new validation:
https://github.com/diennea/herddb/blob/aa7e742d2e9664513745630f94198ca2e788bca0/herddb-core/src/main/java/herddb/codec/RecordSerializer.java#L306

```
if (v == null) {
    if (ColumnTypes.isNotNullDataType(type) {
           throw new StatementExecutionException("Cannot serialize null of a "+typeToString(type)+" column type);
} else {
    return null;
} 
}
```

the same for serializeTo() and for validate() and for serializeValue()

Reply:
Good suggestion.